File: dotnet/picocontainer/src/Test/Tck/AbstractLazyInstantiationTestCase.cs
/*****************************************************************************
 * Copyright (C) PicoContainer Organization. All rights reserved.            *
 * ------------------------------------------------------------------------- *
 * The software in this package is published under the terms of the BSD      *
 * style license a copy of which has been included with this distribution in *
 * the license.txt file.                                                     *
 *                                                                           *
 * Idea by Rachel Davies, Original code by Aslak Hellesoy and Paul Hammant   *
 * C# port by Maarten Grootendorst                                           *
 *****************************************************************************/

using System;
using PicoContainer.Defaults;
using PicoContainer.Tests.TestModel;
using NUnit.Framework;

namespace PicoContainer.Tests.Tck
{
  /// <summary>
  /// Summary description for AbstractLazyInstantiationTestCase.
  /// </summary>
  [TestFixture]
  public abstract class AbstractLazyInstantiationTestCase
  {
    protected abstract IMutablePicoContainer createPicoContainer();

    public class Kilroy 
    {
      public Kilroy(Havana havana) 
      {
        havana.graffiti("Kilroy was here");
      }
    }

    public class Havana 
    {
      public String paint = "Clean wall";

      public void graffiti(String paint) 
      {
        this.paint = paint;
      }
    }

    [Test]
    public void testLazyInstantiation() 
    {
      IMutablePicoContainer pico = createPicoContainer();

      pico.RegisterComponentImplementation(typeof(Kilroy));
      pico.RegisterComponentImplementation(typeof(Havana));

      Assert.AreSame(pico.GetComponentInstance(typeof(Havana)),pico.GetComponentInstance(typeof(Havana)));
      Assert.IsNotNull(pico.GetComponentInstance(typeof(Havana)));
      Assert.AreEqual("Clean wall", ((Havana) pico.GetComponentInstance(typeof(Havana))).paint);
      Assert.IsNotNull(pico.GetComponentInstance(typeof(Kilroy)));
      Assert.AreEqual("Kilroy was here", ((Havana) pico.GetComponentInstance(typeof(Havana))).paint);
    }
  }
}

File: trunk/java/picocontainer/src/test/org/picocontainer/tck/AbstractLazyInstantiationTestCase.java
package org.picocontainer.tck;

import junit.framework.TestCase;
import org.picocontainer.MutablePicoContainer;
import org.picocontainer.PicoException;
import org.picocontainer.PicoRegistrationException;

/**
 * @author Aslak Helles&oslash;y
 * @version $Revision$
 */
public abstract class AbstractLazyInstantiationTestCase extends TestCase {

    protected abstract MutablePicoContainer createPicoContainer();

    public static class Kilroy {
        public Kilroy(Havana havana) {
            havana.graffiti("Kilroy was here");
        }
    }

    public static class Havana {
        public String paint = "Clean wall";

        public void graffiti(String paint) {
            this.paint = paint;
        }
    }

    public void testLazyInstantiation() throws PicoException, PicoRegistrationException {
        MutablePicoContainer pico = createPicoContainer();

        pico.registerComponentImplementation(Kilroy.class);
        pico.registerComponentImplementation(Havana.class);

        assertSame(pico.getComponentInstance(Havana.class), pico.getComponentInstance(Havana.class));
        assertNotNull(pico.getComponentInstance(Havana.class));
        assertEquals("Clean wall", ((Havana) pico.getComponentInstance(Havana.class)).paint);
        assertNotNull(pico.getComponentInstance(Kilroy.class));
        assertEquals("Kilroy was here", ((Havana) pico.getComponentInstance(Havana.class)).paint);
    }
}
